shopfloor/services/delivery.py

```diff
@@ -54,21 +135,200 @@ def scan_deliver(self, barcode):
+            if len(packages.mapped("quant_ids.product_id")) > 1:
+                return self._response_for_deliver(
+                    new_picking,
+                    message=self.msg_store.product_mixed_package_scan_package(),
+                )
+        self._set_lines_done(lines)
+        return self._response_for_deliver(new_picking)
+
+    def _deliver_lot(self, picking, lot):
+        lines = self.env["stock.move.line"].search(self._lines_from_lot_domain(lot))
+        if not lines:
+            return self._response_for_deliver(
+                picking, message=self.msg_store.lot_not_found_in_pickings()
+            )
+
+        new_picking = fields.first(lines.mapped("picking_id"))
+
+        # When lots are as units outside of packages, we can select them for
+        # packing, but if they are in a package, we want the user to scan the packages.
+        # If the product is only in one package though, scanning the lot selects
+        # the package.
+        packages = lines.mapped("package_id")
+        # Do not use mapped here: we want to see if we have more than one
+        # package, but also if we have one lot as a package and the same lot as
+        # a unit in another line. In both cases, we want the user to scan the
+        # package.
+        if packages and len({l.package_id for l in lines}) > 1:
+            return self._response_for_deliver(
+                new_picking, message=self.msg_store.lot_multiple_packages_scan_package()
+            )
+        elif packages:
+            # we have 1 package
+            # abort the operation if the package contain more than one product
+            if len(packages.quant_ids) > 1:
+                return self._response_for_deliver(
+                    new_picking,
+                    message=self.msg_store.lot_mixed_package_scan_package(),
+                )
+
+        self._set_lines_done(lines)
+        return self._response_for_deliver(new_picking)
+
+    def _action_picking_done(self, picking):
+        """Try to validate the stock picking if all its lines have been processed.
+
+        Return `True` if the picking has been validated successfully.
+        """
+        move_lines_done = all(
+            [line.qty_done >= line.product_uom_qty for line in picking.move_line_ids]
+        )
+        if move_lines_done:
+            picking.action_done()
+            return True
+        return False
 
     def list_stock_picking(self):
-        """Return the list of stock pickings for the picking type
+        """Return the list of stock pickings for the picking types
 
         It returns only stock picking available or partially available.
 
         Transitions:
         * manual_selection: next state to show the list of stock pickings
         """
-        return self._response()
+        pickings = self.env["stock.picking"].search(self._pickings_domain(), order="id")
+        return self._response_for_manual_selection(pickings)
```

Comment:
Should we include stats for the completion rate of a transfer? Or do you compute it on the client @simahawk?

Reply:
I do it client side. In any case we'll have to refactor something because we'll have to handle the state by truck which we cannot do now. But that's another story :)